pyomo/contrib/mindtpy/cut_generation.py

```diff
@@ -78,129 +43,57 @@ def add_oa_cuts(target_model, dual_values, solve_data, config,
-    dual_values:
-        contains the value of the duals for each constraint
-    solve_data: MindtPy Data Container
-        data container that holds solve-instance data
-    config: ConfigBlock
-        contains the specific configurations for the algorithm
-    linearize_active: bool, optional
-        this parameter acts as a Boolean flag that signals whether the linearized constraint is active
-    linearize_violated: bool, optional
-        this parameter acts as a Boolean flag that signals whether the nonlinear constraint represented by the
-        linearized constraint has been violated
-    """
-    for index, constr in enumerate(target_model.MindtPy_utils.constraint_list):
-        if constr.body.polynomial_degree() in (0, 1):
-            continue
-
-        constr_vars = list(identify_variables(constr.body))
-        jacs = solve_data.jacobians
-
-        # Equality constraint (makes the problem nonconvex)
-        if constr.has_ub() and constr.has_lb() and constr.upper == constr.lower and config.use_dual:
-            sign_adjust = -1 if solve_data.objective_sense == minimize else 1
-            rhs = constr.lower
-            if config.add_slack:
-                slack_var = target_model.MindtPy_utils.MindtPy_linear_cuts.slack_vars.add()
-            target_model.MindtPy_utils.MindtPy_linear_cuts.oa_cuts.add(
-                expr=copysign(1, sign_adjust * dual_values[index])
-                * (sum(value(jacs[constr][var]) * (var - value(var))
-                       for var in list(EXPR.identify_variables(constr.body)))
-                    + value(constr.body) - rhs)
-                - (slack_var if config.add_slack else 0) <= 0)
-
-        else:  # Inequality constraint (possibly two-sided)
-            if constr.has_ub() \
-                and (linearize_active and abs(constr.uslack()) < config.zero_tolerance) \
-                    or (linearize_violated and constr.uslack() < 0) \
-                    or (config.linearize_inactive and constr.uslack() > 0):
-                if config.add_slack:
-                    slack_var = target_model.MindtPy_utils.MindtPy_linear_cuts.slack_vars.add()
-
-                target_model.MindtPy_utils.MindtPy_linear_cuts.oa_cuts.add(
-                    expr=(sum(value(jacs[constr][var])*(var - var.value)
-                              for var in constr_vars) + value(constr.body)
-                          - (slack_var if config.add_slack else 0)
-                          <= constr.upper)
-                )
-
-            if constr.has_lb() \
-                and (linearize_active and abs(constr.lslack()) < config.zero_tolerance) \
-                    or (linearize_violated and constr.lslack() < 0) \
-                    or (config.linearize_inactive and constr.lslack() > 0):
+            # Equality constraint (makes the problem nonconvex)
+            if constr.has_ub() and constr.has_lb() and constr.upper == constr.lower and config.equality_relaxation:
```

Comment:
I do not quite understand how `lower` or `upper` contain a mutable Param. Should I also add `value()` function for `rhs = constr.lower`.

Thanks.

Reply:
@ZedongPeng The "trick" with mutable Params is that they are preserved in the expression tree (unlike immutable Params, which are immediately resolved to their numeric values).  Pyomo reformulates constraints (equality and inequality) so that all "potentially variable" terms are moved to the constraint body, but non-potentially variable terms (i.e., anything that cannot contain a variable) is left in the upper / lower bounds.  For example:
```python
>>> from pyomo.environ import *
>>> m = ConcreteModel()
>>> m.x = Var()
>>> m.p = Param(mutable=True)
>>> m.c = Constraint(expr=m.x <= m.p)
>>> type(m.c.upper)
<class 'pyomo.core.base.param.SimpleParam'>
```
In addition, upper / lower can also be (non-potentially variable) expressions:
```
>>> m.d = Constraint(expr=m.x**2 >= 1 - m.p)
>>> m.d.lower
<pyomo.core.expr.numeric_expr.NPV_SumExpression object at 0x7f76e44a2150>
>>> str(m.d.lower)
'1 - p'
```